### ISSUE TITLE
Compiled SPIR-V fails Vulkan validation due to type mismatch

### ISSUE THREAD

**Issue by jwollen** [KhronosGroup/glslang#307]
The attached shader compiles without error to SPIR-V, however fails validation when fed to Vulkan with
`Error: SPIR-V module not valid: OpStore Pointer <id> '212's type does not match Object <id> '199's type. ([5] SC)` (or causes crashes without validation).

The referenced type is the `PointLightData` structure and it's usage as a parameter to `ProcessLight_id53`. Similar shaders with uniforms of structure types seem to work fine, when not passed to functions.

Tested with the Vulkan SDK 1.0.13 versions of glslang and layers.

Possibly duplicate of #304. 

[PointLight.frag.txt](https://github.com/KhronosGroup/glslang/files/278663/PointLight.frag.txt)
[PointLight.spv.txt](https://github.com/KhronosGroup/glslang/files/278665/PointLight.spv.txt)

**Comment by johnkslang** [comment]
It is similar to issue #304: The two struct types exist due to possible two sets of decorations for two types of layouts.

The question is, which layer is responsible for specialization of functions that have the same type in GLSL, but different types in SPIR-V.

Imagine the following GLSL-like language that can pass by reference:

```
struct S { mat4 m; }
uniform blockName { S s1; }; // need an S with decoration
S s2;  // no decorations on S

void foo(struct S& s); // s could be laid out two different ways

void main()
{
  foo(s1);
  foo(s2);
}
```

At a GLSL level, there is one `foo()`, while at a lower level, may need a `fooTemp()` for an internally laid out matrix and a `fooUniform()` for a transparently laid out matrix.

**Comment by twanks123** [comment]
Hello, i have the same problem and fixed it by doing this:

for(int i = 0; i < lightCounts.NUM_DIRECTIONAL_LIGHTS; i++)
	{
		// structs passed to this shader are different from the structs passed to the functions
		vec3 color     = lighting.directionalLights[i].base.color;
		float intensity = lighting.directionalLights[i].base.intensity;
		vec3 dir 	      = lighting.directionalLights[i].direction;
		
		DirectionalLight dirLight = DirectionalLight(BaseLight(color, intensity), dir);
		
		totalLight += calcDirectionalLight(dirLight, normal);
	}	

instead of this:
for(int i = 0; i < lightCounts.NUM_DIRECTIONAL_LIGHTS; i++)
	{
               totalLight += calcDirectionalLight(lighting.directionalLights[i], normal);
       }

But this "hack" looks so ugly and i dont know how to fix it properly. Any suggestions for me?

**Comment by cippyboy** [comment]
I have the same issue with my shader too :

ERROR: [SC] Code 5 : SPIR-V module not valid: OpStore Pointer <id> '321's type does not match Object <id> '314's type

```
#version 310 es
#define HasDiffuseTexture 1
#define HasVertices 1
#define HasTexcoords 1
#define HasNormals 1
#line 2
precision highp float;
precision highp int;

layout(location = 0) out vec4 FragColor;

#ifdef NormalsRT
	layout(location = NormalsRTSlot ) out vec4 NormalRT;
#endif
#ifdef DepthRT	
	layout(location = 1) out vec4 DepthOut;
#endif
#if defined VelocityRT
	layout(location = VelocityRTSlot) out uvec2 VelocityOut;
#endif

uniform highp sampler2DShadow ShadowMap;

struct Light
{
	vec3 Position;
	int LightType;
	
	vec3 Direction;
	vec3 Attenuation;
	
	vec3 Ambient;
	vec3 Diffuse;
	vec3 Specular;	
	float Intensity;
};

layout(std140, binding = 1, set = 0 ) uniform PerFrameBuffer
{		
	uniform mat4 ProjectionMatrix;
	uniform mat4 ViewMatrix;
	uniform mat4 WorldMatrix;
	uniform mat4 InverseWorldMatrix;
	uniform mat4 ShadowViewMatrix;
	uniform Light Light0;
	uniform vec4 MaterialColor;
};

layout( location = 0 ) in vec3 WorldPos;

#if defined HasTexcoords 
	layout( location = 1 ) in vec2 OutTexcoord;
#endif

#if defined HasNormals 
	layout( location = 2 ) in vec3 OutNormal;
#endif

#if defined VelocityRT
	layout( location = 3 ) in vec2 VelocityVertex;
#endif

uniform sampler2D Texture0;

void ComputeLight( Light LightProperties, vec3 WorldPosition, vec3 Normal, vec3 WorldViewVector,
                  inout vec3 TotalAmbient, inout vec3 TotalDiffuse, inout vec3 TotalSpecular)
{
	//Stupid GL camera transform
	vec3 Position = LightProperties.Position;
	//Position.z *=-1;
    
	vec3 LightDirection = Position - WorldPosition;
	float DistanceToLight = length( LightDirection );
	float DistanceToLightSquared = dot(LightDirection, LightDirection);
	vec3 Att = LightProperties.Attenuation;
	float Attenuation = Att.x + Att.y * DistanceToLight + Att.z * DistanceToLightSquared;
	Attenuation = max( 1.0, Attenuation);
	Attenuation = 1.0 / Attenuation;	
    
	LightDirection = normalize( LightDirection );
	float dotres = dot( Normal, LightDirection );
    vec3 NDotL = vec3(dotres,dotres,dotres);
	float Contribution = max( 0.0, dotres );
	
	float SpecularContribution = 0.0;
	/*if (0)
	{
		float SpecularMultiplier = 0.0;
        if (dotres > 0.0)
            SpecularMultiplier = 1.0;
        
		vec3 LightHalfVector = LightDirection + WorldViewVector;
        LightHalfVector = normalize( LightHalfVector );               
        
        float LightNDotH = max(0.0, dot(Normal, LightHalfVector) );
        SpecularContribution = pow( LightNDotH, 64.0 );                
        SpecularContribution = SpecularContribution * SpecularMultiplier * Attenuation;
    }*/
	
	TotalAmbient  += LightProperties.Ambient;
	TotalDiffuse  += Contribution * LightProperties.Diffuse * LightProperties.Intensity;// * Attenuation;
	TotalSpecular += SpecularContribution * LightProperties.Specular;// * Attenuation;
}
float textureCompare(sampler2D depths, vec2 uv, float compare)
{
    float depth = texture(depths, uv).r;
    return step(compare, depth);
}
float IsShadowed( vec2 Texcoord, float LightDepth, vec2 Offset, vec2 ShadowMapSizeInverse )
{
	vec2 TexOffset = vec2( Offset.x * ShadowMapSizeInverse.x, Offset.y * ShadowMapSizeInverse.y );	
	#ifdef ShadowSamplers
		float Depth = texture( ShadowMap, vec3( Texcoord.x + TexOffset.x,  Texcoord.y + TexOffset.y, LightDepth)  );
		return 1.0 - Depth;
	#elif defined EnableShadows
		vec2 TexelProj = Texcoord * ShadowMapSize + 0.5 ;
		vec2 f = fract( TexelProj );
		vec2 centroidUV = floor( TexelProj ) / ShadowMapSize;

		float Depth00 = textureCompare( ShadowMap, centroidUV + TexOffset, LightDepth);
		float Depth01 = textureCompare( ShadowMap, centroidUV + TexOffset + vec2(0.0, ShadowMapSizeInverse.y ) , LightDepth);		
		float Depth10 = textureCompare( ShadowMap, centroidUV + TexOffset + vec2(ShadowMapSizeInverse.x, 0.0), LightDepth);
		float Depth11 = textureCompare( ShadowMap, centroidUV + TexOffset + ShadowMapSizeInverse, LightDepth );

		float L1 = mix( Depth00, Depth01, f.y);
	    float L2 = mix( Depth10, Depth11, f.y);
	    
		float Depth = mix( L1, L2, f.x);
	
		return 1.0 - Depth;		
	#else
		return 1.0;
	#endif
}
float GetFilteredShadow(vec2 Texcoord, vec2 ShadowMapSizeInverse, float LightSpaceDepth, float GridSize)
{	
	float Shadowed = 0.0;
	float Shadow = 1.0;
	float coef = 1.0;
					
	//float GridSize = ShadowGridSize;//1=3x3,2=5x5
	
	for(float x=-GridSize; x<=GridSize; x++)
		for(float y=-GridSize; y<=GridSize; y++)
		{
			vec2 Offset = vec2( x, y );
			Shadowed += IsShadowed( Texcoord, LightSpaceDepth, Offset, ShadowMapSizeInverse );
		}
	
	coef = (GridSize*2.0+1.0) * (GridSize*2.0+1.0);

	if (Shadowed > 0.0)
	{
		Shadow = 1.0 - Shadowed/coef;
		//Shadow = 1.0 - Shadowed;
	}

	return Shadow;
}
float ComputeShadow(vec2 ShadowTextureSize, float ShadowBias, vec3 WorldPos, mat4 ShadowViewMatrix, float GridSize)
{
	//Optimization, move this to vertex shader !
	vec4 LightProjPos = ShadowViewMatrix * vec4(WorldPos.xyz,1);

	vec3 LightDiv = vec3( LightProjPos.x / LightProjPos.w, LightProjPos.y / LightProjPos.w, LightProjPos.z / LightProjPos.w);
	vec3 ShadowTexC = vec3( 0.5, 0.5, 0.5 );
	ShadowTexC *= LightDiv;
	ShadowTexC += vec3( 0.5, 0.5, 0.5 );

	//Flip Y, only for DX
	//ShadowTexC.y = 1.0 - ShadowTexC.y;

	//float LightSpaceDepth = (LightProjPos.z / LightProjPos.w);//DX Style ?
	float LightSpaceDepth = ShadowTexC.z - ShadowBias;
		
	//float Depth = texture( ShadowMap, ShadowTexC.xy ).x;

    if ( LightProjPos.z < 0.0 || LightProjPos.z > 500.0 )
		//|| LightProjPos.x < 0.0 || LightProjPos.x > 1.0
		//|| LightProjPos.y < 0.0 || LightProjPos.y > 1.0)
    {
		return 1.0;//lit
    }
	else
	{
		vec2 ShadowMapSizeInverse = 1.0 / ShadowTextureSize;
		return GetFilteredShadow( ShadowTexC.xy, ShadowMapSizeInverse, LightSpaceDepth, GridSize );
	}
}
void main()
{
	#ifdef EnableCullHeight
	
		if ( WorldPos.y < CullHeight)
			discard;
	
	#endif

    vec3 Normal = vec3(0.0,0.0,1.0);
	vec3 WorldNormal;
    #if defined HasNormals 
		//this is already done in VS
		//WorldNormal = OutNormal * InverseWorldMatrix;
        Normal = normalize( OutNormal );
    #endif
    
    vec3 TotalAmbient = vec3(0.0,0.0,0.0);	
	vec3 TotalDiffuse = vec3(0.0,0.0,0.0);	
	vec3 TotalSpecular = vec3(0.0,0.0,0.0);	

	vec3 SpecularColor = vec3(0.0,0.0,0.0);	
    
    vec3 WorldViewVector = vec3(0.0,-1.0,0.0);//CameraPosition - WorldPos;
	WorldViewVector = normalize( WorldViewVector );
    
    #if defined HasNormals
        ComputeLight( Light0, WorldPos, Normal, WorldViewVector, TotalAmbient, TotalDiffuse, TotalSpecular );
    #endif
    
    vec3 DiffuseTexture = vec3(0.5, 0.5, 0.5 );
    
    #if defined(HasDiffuseTexture) && defined(HasTexcoords)
		DiffuseTexture = texture( Texture0, OutTexcoord ).rgb;
    #endif
    
    float Alpha = MaterialColor.a;
    
    #ifdef EnableAlphaTesting
        float Average = (DiffuseTexture.r + DiffuseTexture.g + DiffuseTexture.b)/3.0;
        if (Average > 0.8)
            discard;
    
        Alpha = mix( 1.0, 0.0, Average );
	#elif defined DynamicShaderLinking
		DoAlphaTestingUniform( DiffuseTexture, Alpha );
    #endif
    
     vec3 FinalColor = vec3(0.0, 0.0, 0.0);
    
	bool UseShadows = false;
	float Shadow = 1.0;
	vec2 ShadowMapSizeCopy;
	#ifdef EnableShadows
		ShadowMapSizeCopy = ShadowMapSize;//textureSize( ShadowMap, 0 );//
		UseShadows = true;
	
	if (UseShadows)
	{
		float ShadowBias = 0.0;//0.0005;
		Shadow = ComputeShadow( ShadowMapSizeCopy, ShadowBias, WorldPos, ShadowViewMatrix, ShadowGridSize );
	}

	#endif

    #if defined HasNormals 
        //FinalColor = TotalAmbient*0.3 + (TotalDiffuse + SpecularColor* 10.0 * TotalSpecular);
		FinalColor = TotalAmbient + (TotalDiffuse + SpecularColor* 10.0 * TotalSpecular) * Shadow;
    #else
        FinalColor = vec3(1.0, 1.0, 1.0);
    #endif
	
    #if defined HasDiffuseTexture 
        FinalColor *= DiffuseTexture;
	#else
        FinalColor *= MaterialColor.rgb;
    #endif

	//Debug->
	//FinalColor = DiffuseTexture.rgb;
	//<-

    FragColor = vec4( FinalColor , Alpha );

	#ifdef NormalsRT
		//vec3 ViewNormal = mat3(ViewMatrix) * Normal;
		NormalRT = vec4( Normal * 0.5 + 0.5, 1.0 );
	#endif
	#ifdef DepthRT
		DepthOut = gl_FragCoord.z;
	#endif
	#if defined VelocityRT		
		VelocityOut = uvec2( vec2( VelocityVertex.xy * vec2(0.5) + vec2(0.5) ) * vec2( 65535.0 ) );
	#endif
}
```

**Comment by johnkslang** [comment]
Closing as a duplicate of #304, so there is just one place to discuss.  But it is indeed important, it has shown up from multiple directions.

**Comment by johnkslang** [comment]
Reopening, as the point of issue #304 is now fixed by 4bf7155, while r-value parameter handling is not yet complete.

This commit does, however, do recursive flattening for structure passing to functions, where the l-value is passed by the call site.  That basically leaves out "const in" parameters for structure types that are aliased.

**Comment by johnkslang** [comment]
This case currently validates cleanly, with glslang doing member-wise copies. Closing.
